jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py

```diff
@@ -159,47 +199,147 @@ def _on_spectral_unit_changed(self, *args):
         if self.app.config == 'specviz':
             return
 
-        # Check for a scale factor/data passed through spectral extraction plugin.
-        specs_w_factor = [spec for spec in self.app.data_collection
-                          if "_pixel_scale_factor" in spec.meta]
-        # Translate if we have a scale factor
-        if specs_w_factor:
-            self.translate_units(self.flux_or_sb_selected)
-        # The translator dropdown hasn't been loaded yet so don't try translating
-        elif not self.show_translator:
+        # we want to raise an error if a user tries to translate with an
+        # untranslated Flux unit using the API
+        untranslatable_units = self.untranslatable_units()
+        untranslatable_units = units_to_strings(untranslatable_units)
+
+        if hasattr(self, 'flux_unit'):
+            if (self.flux_unit.selected in untranslatable_units) \
+                 and (flux_or_sb == 'Surface Brightness'):
+                raise ValueError(
+                    f"Selected flux unit is not translatable. Please choose a flux unit "
+                    f"that is not in the following list: {untranslatable_units}."
+                )
+
+        if self.spectrum_viewer.state.y_display_unit:
+            spec_units = u.Unit(self.spectrum_viewer.state.y_display_unit)
+        else:
             return
-        # Notify the user to extract a spectrum before using the surface brightness/flux
-        # translation. Can be removed after all 1D spectra in Cubeviz pass through
-        # spectral extraction plugin (as the scale factor will then be stored).
+        # on instantiation, we set determine flux choices and selection
+        # after surface brightness
+        if not self.flux_unit.choices:
+            return
+        # Surface Brightness -> Flux
+        if check_if_unit_is_per_solid_angle(spec_units) and flux_or_sb == 'Flux':
+            spec_units *= u.sr
+            # update display units
+            self.spectrum_viewer.state.y_display_unit = str(spec_units)
+            self.flux_or_sb.selected = 'Flux'
+
+        # Flux -> Surface Brightness
+        elif (not check_if_unit_is_per_solid_angle(spec_units)
+              and flux_or_sb == 'Surface Brightness'):
+            spec_units /= u.sr
+            # update display units
+            self.spectrum_viewer.state.y_display_unit = str(spec_units)
+            self.flux_or_sb.selected = 'Surface Brightness'
+        # entered the translator when we shouldn't translate
         else:
-            raise ValueError("No collapsed spectra in data collection, \
-                              please collapse a spectrum first.")
+            return
+
+        self.hub.broadcast(GlobalDisplayUnitChanged('flux',
+                                                    spec_units,
+                                                    sender=self))
+        self.spectrum_viewer.reset_limits()
+
+    def untranslatable_units(self):
```

Comment:
should this be a property since it doesn't take arguments, or maybe even a cached property so it doesn't have to be re-built each time its called in `_translate`?

Reply:
I believe I made this adjustment, let me know if this is correct or I need to revise